### PR TITLE
add DEP

### PR DESCRIPTION
**Pull request by esborisova**
1. In the paper they mention that punctuation was excluded. But they do not specify how and at which stage. For this reason, the script still considers all the punctuation.
2. in [this line](https://github.com/ryabhmd/scilons-eval/blob/e463d2e783750b7feec6971931b161e59c7c46c2/model_eval.py#L137), the path to the test data is required or maybe all the tokens from the test set can be collected earlier/in a different way. 
3. the scores for DEP are way lower than in a paper, I could not figure out the issue.
4. we can use data collator for padding and remove lines 243-246 [here](https://github.com/ryabhmd/scilons-eval/blob/e463d2e783750b7feec6971931b161e59c7c46c2/data/data_prep_utils.py#L243).

### CODE REVIEW

File: data/prepare_data.py [hunk 1]
```diff
@@ -70,8 +86,45 @@ def run(self) -> (DatasetDict, Dict):
...
+                    }
+                )
+
+                dataset_dict["dep"] = dataset_inputs_dep
+                dataset_dict["heads"] = dataset_inputs_heads
+
+                dataset_dict[dataset_name[:-4]] = {"dep": dataset_inputs_dep, "heads": dataset_inputs_heads}
+
+
             dataset_inputs = Dataset.from_dict(
```
Comment (ryabhmd) [inline data/prepare_data.py:128]: Lines from here forward should be executed only if the task is not dep, so they should be in an else statement, I think. 
Alternatively, the second if statement in line 139 can be moved to be inside this if.

File: data/eva_model_utils.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import torch
```
Comment (ryabhmd) [inline data/eva_model_utils.py:1]: A general comment on this file is that I think it should be renamed to 'dep_eval_utils.py' or something like this, because it is only relevant for dep. 
Also, I think it should be outside of the 'data' dir, since it is connected to evaluation and not data preparation.

File: model_eval.py [hunk 3]
```diff
@@ -37,6 +38,10 @@ def train_model(self):
             device=self.device,
             max_length = self.max_length
         )
+        
+        if self.task == "dep":
+            dataset_dict, labels_mapper_dep, labels_mapper_head = dataset_prep.run()
+
         dataset_dict, labels_mapper = dataset_prep.run()
```
Comment (ryabhmd) [inline model_eval.py:45]: This should be in an else statement, because otherwise it will overrun the 'dataset_dict' you have inside the if above, no?

File: model_eval.py [hunk 4]
```diff
@@ -60,6 +72,28 @@ def train_model(self):
...
+                )
+            
+            trainer_dep.train()
+            trainer_dep.save_model("results/models")
+
+            trainer_heads.train()
+            trainer_heads.save_model("results/models")
+
+
         trainer = Trainer(
```
Comment (ryabhmd) [inline model_eval.py:97]: this trainer should also be in an else statement.

File: model_eval.py [hunk 5]
```diff
@@ -71,9 +105,44 @@ def train_model(self):
 
         trainer.save_model("results/models")
 
+        if self.task == "dep":
```
Comment (ryabhmd) [inline model_eval.py:108]: This can be at the end of the previous if statement (line 75)

File: model_eval.py [hunk 6]
```diff
@@ -170,7 +239,7 @@ def evaluate_model(self):
         # Placeholder for DEP code
         #elif self.task == 'dep':
 
-        return micro_f1, macro_f1
+        return micro_f1, macro_f1, uas, las
```
Comment (ryabhmd) [inline model_eval.py:242]: This will throw an error for any task other than dep, so should be inside the if that has all the dep model training inside it.

File: model_eval.py [hunk 7]
```diff
@@ -71,9 +105,44 @@ def train_model(self):
...
+            words_test_set = extract_dep_data("data/genia/test.txt") # need to define the correct path
+            words_test_set = [item[0] for sample in words_test_set for item in sample]
+
+            uas, las = calculate_uas_las(words_test_set, 
+                                         gold_head_labels, 
+                                         gold_dep_labels, 
+                                         pred_head_labels, 
+                                         pred_dep_labels)
+
         trained_model, dataset_dict, labels_mapper = self.train_model()
```
Comment (ryabhmd) [inline model_eval.py:146]: Everything below here should be in an else statement unless a return statement is located at the end of the previous if statement in line 114.

File: model_eval.py [hunk 8]
```diff
@@ -218,17 +287,18 @@ def main():
         max_length = max_length
     )
 
-    micro_f1, macro_f1 = model_eval.evaluate_model()
+    micro_f1, macro_f1, uas, las = model_eval.evaluate_model()
```
Comment (ryabhmd) [inline model_eval.py:290]: Again, this will throw an error for anything other than dep.

**Review by ryabhmd — COMMENTED** [review]
Reviewed the code and will work on merging it with the other tasks (at some places the placement of if/else statements should be corrected). 
After this is done, I will take a closer look at the training/evaluation code to see if I can understand the difference between our scores and SciBERT.
